src/sinol_make/commands/run/__init__.py

```diff
@@ -194,50 +196,114 @@ def compile(self, solution):
-		if code == 35072:
-			result["Status"] = "TL"
-		elif "Time" in result.keys() and result["Time"] > time_limit:
-			result["Status"] = "TL"
-		elif "Memory" in result.keys() and result["Memory"] > memory_limit:
-			result["Status"] = "ML"
-		elif "Status" not in result.keys():
-			result["Status"] = "RE"
-		elif result["Status"] == "OK":
-			if result["Time"] > time_limit:
-				result["Status"] = "TL"
-			elif result["Memory"] > memory_limit:
-				result["Status"] = "ML"
+					if key == "Time":
+						result.Time = self.parse_time(value)
+					elif key == "Memory":
+						result.Memory = self.parse_memory(value)
+					else:
+						setattr(result, key, value)
+
+		if timeout_exit_code == 35072:
+			result.Status = "TL"
+		elif getattr(result, "Time") is not None and result.Time > time_limit:
+			result.Status = "TL"
+		elif getattr(result, "Memory") is not None and result.Memory > memory_limit:
+			result.Status = "ML"
+		elif getattr(result, "Status") is None:
+			result.Status = "RE"
+		elif result.Status == "OK":
+			if result.Time > time_limit:
+				result.Status = "TL"
+			elif result.Memory > memory_limit:
+				result.Status = "ML"
 			elif os.system("diff -q -Z %s %s >/dev/null"
-						% (output_file, self.get_output_file(test))):
-				result["Status"] = "WA"
+						% (output_file, answer_file)):
+				result.Status = "WA"
+		else:
+			result.Status = result.Status[:2]
+
+		return result
+
+
+	def execute_time(self, command, result_file, output_file, answer_file, time_limit, memory_limit):
+		timeout_exit_code = os.system(command)
+
+		result = ExecutionResult(None, None, None)
+		lines = open(result_file).readlines()
+		program_exit_code = None
+		if len(lines) == 3:
+			"""
+			If programs runs successfully, the output looks like this:
+			 - first line is CPU time in seconds
+			 - second line is memory in Kbytes
+			 - third line is exit code
+			This format is defined by -f flag in time command.
+			"""
+			result.Time = round(float(lines[0].strip()) * 1000)
+			result.Memory = int(lines[1].strip())
+			program_exit_code = int(lines[2].strip())
+		if len(lines) > 0 and "Command terminated by signal " in lines[0]:
+			"""
+			If there was a runtime error, the first line is the error message with signal number.
+			For example:
+				Command terminated by signal 11
+			"""
+			program_exit_code = int(lines[0].strip().split(" ")[-1])
+
+		if program_exit_code != None and program_exit_code != 0:
+			result.Status = "RE"
+		elif timeout_exit_code != 0:
+			result.Status = "TL"
+		elif result.Time > time_limit:
+			result.Status = "TL"
+		elif result.Memory > memory_limit:
+			result.Status = "ML"
+		elif os.system("diff -q -Z %s %s >/dev/null" % (output_file, answer_file)):
+			result.Status = "WA"
 		else:
-			result["Status"] = result["Status"][:2]
+			result.Status = "OK"
+
 		return result
 
+
+	def execute(self, execution):
+		(name, executable, test, time_limit, memory_limit, timetool_path) = execution
```

Comment:
The function name and argument name are quite similar and thus I'm not sure what the function exactly does based on its signature. It's also quite similar to the function name `perform_executions`, which also confused me. Let's find more intuitive names to the functions/arguments.

Reply:
I changed names of the fuctions 7215e30ccd3d94fe1e84eb4080a27ef9102abfaa